RC4: java.lang.NullPointerException in saveFileWithParent()

> java.lang.RuntimeException: 
  at android.os.AsyncTask$3.done (AsyncTask.java:325)
  at java.util.concurrent.FutureTask.finishCompletion (FutureTask.java:354)
  at java.util.concurrent.FutureTask.setException (FutureTask.java:223)
  at java.util.concurrent.FutureTask.run (FutureTask.java:242)
  at android.os.AsyncTask$SerialExecutor$1.run (AsyncTask.java:243)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1133)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:607)
  at java.lang.Thread.run (Thread.java:761)
Caused by: java.lang.NullPointerException: 
  at com.owncloud.android.datamodel.FileDataStorageManager.saveFileWithParent (FileDataStorageManager.java:277)
  at com.owncloud.android.datamodel.FileDataStorageManager.saveFileWithParent (FileDataStorageManager.java:268)
  at com.owncloud.android.ui.adapter.FileListListAdapter.parseVirtuals (FileListListAdapter.java:590)
  at com.owncloud.android.ui.adapter.FileListListAdapter.setData (FileListListAdapter.java:504)
  at com.owncloud.android.ui.fragment.OCFileListFragment$12.doInBackground (OCFileListFragment.java:1526)
  at android.os.AsyncTask$2.call (AsyncTask.java:305)
  at java.util.concurrent.FutureTask.run (FutureTask.java:237)

For some reason the root ("/") OcFile is not stored in the database and thus the NPE occurs.
Solution: if root file does not exist, fetch it from server.

--- a/src/gplay/AndroidManifest.xml
+++ b/src/gplay/AndroidManifest.xml
@@ -19,8 +19,8 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           xmlns:tools="http://schemas.android.com/tools"
           package="com.owncloud.android"
-          android:versionCode="20000004"
-          android:versionName="2.0.0RC4">
+          android:versionCode="20000005"
+          android:versionName="2.0.0RC5">
 
     <application
         android:name=".MainApp"

--- a/src/main/AndroidManifest.xml
+++ b/src/main/AndroidManifest.xml
@@ -20,8 +20,8 @@
 -->
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           package="com.owncloud.android"
-          android:versionCode="20000004"
-          android:versionName="2.0.0RC4">
+          android:versionCode="20000005"
+          android:versionName="2.0.0RC5">
 
     <uses-sdk
         android:minSdkVersion="14"

--- a/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java
+++ b/src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java
@@ -44,6 +44,7 @@ import com.owncloud.android.lib.resources.shares.OCShare;
 import com.owncloud.android.lib.resources.shares.ShareType;
 import com.owncloud.android.lib.resources.status.CapabilityBooleanType;
 import com.owncloud.android.lib.resources.status.OCCapability;
+import com.owncloud.android.operations.RemoteOperationFailedException;
 import com.owncloud.android.utils.FileStorageUtils;
 import com.owncloud.android.utils.MimeType;
 import com.owncloud.android.utils.MimeTypeUtil;
@@ -250,6 +251,14 @@ public class FileDataStorageManager {
         return overriden;
     }
 
+    /**
+     * traverses a files parent tree to be able to store a file with its parents.
+     * Throws a RemoteOperationFailedException in case the parent can't be retrieved.
+     *
+     * @param file the file
+     * @param context the app context
+     * @return the parent file
+     */
     public OCFile saveFileWithParent(OCFile file, Context context) {
         if (file.getParentId() == 0 && !file.getRemotePath().equals("/")) {
             String remotePath = file.getRemotePath();
@@ -267,8 +276,15 @@ public class FileDataStorageManager {
 
                     returnFile = saveFileWithParent(remoteFolder, context);
                 } else {
-                    returnFile = null;
-                    Log_OC.e(TAG, "Error during saving file with parents: " + file.getRemotePath());
+                    Exception exception = result.getException();
+                    String message = "Error during saving file with parents: " + file.getRemotePath() + " / "
+                            + result.getLogMessage();
+
+                    if (exception != null) {
+                        throw new RemoteOperationFailedException(message, exception);
+                    } else {
+                        throw new RemoteOperationFailedException(message);
+                    }
                 }
             } else {
                 returnFile = saveFileWithParent(parentFile, context);

--- a/src/main/java/com/owncloud/android/operations/RemoteOperationFailedException.java
+++ b/src/main/java/com/owncloud/android/operations/RemoteOperationFailedException.java
@@ -1,0 +1,32 @@
+package com.owncloud.android.operations;
+
+/**
+ * RuntimeException for throwing errors of remote operation calls.
+ */
+public class RemoteOperationFailedException extends RuntimeException {
+    /**
+     * Constructs a new runtime exception with the specified detail message and
+     * cause.
+     *
+     * @param message the detail message (which is saved for later retrieval
+     *                by the {@link #getMessage()} method).
+     * @param cause   the cause (which is saved for later retrieval by the
+     *                {@link #getCause()} method).  (A <code>null</code> value
+     *                is permitted, and indicates that the cause is nonexistent
+     *                or unknown.)
+     */
+    public RemoteOperationFailedException(String message, Throwable cause) {
+        super(message + " / " + cause.getMessage(), cause);
+    }
+
+    /**
+     * Constructs a new runtime exception with the specified detail message and
+     * cause.
+     *
+     * @param message the detail message (which is saved for later retrieval
+     *                by the {@link #getMessage()} method).
+     */
+    public RemoteOperationFailedException(String message) {
+        super(message);
+    }
+}

--- a/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java
+++ b/src/main/java/com/owncloud/android/ui/adapter/FileListListAdapter.java
@@ -59,6 +59,7 @@ import com.owncloud.android.lib.resources.files.ReadRemoteFileOperation;
 import com.owncloud.android.lib.resources.files.RemoteFile;
 import com.owncloud.android.lib.resources.shares.OCShare;
 import com.owncloud.android.lib.resources.shares.ShareType;
+import com.owncloud.android.operations.RemoteOperationFailedException;
 import com.owncloud.android.services.OperationsService.OperationsServiceBinder;
 import com.owncloud.android.ui.activity.ComponentsGetter;
 import com.owncloud.android.ui.activity.FileActivity;
@@ -587,17 +588,22 @@ public class FileListListAdapter extends BaseAdapter {
         for (int i = 0; i < objects.size(); i++) {
             OCFile ocFile = FileStorageUtils.fillOCFile((RemoteFile) objects.get(i));
             searchForLocalFileInDefaultPath(ocFile);
-            ocFile = mStorageManager.saveFileWithParent(ocFile, mContext);
 
-            if (!onlyImages || MimeTypeUtil.isImage(ocFile)) {
-                mFiles.add(ocFile);
+            try {
+                ocFile = mStorageManager.saveFileWithParent(ocFile, mContext);
+
+                if (!onlyImages || MimeTypeUtil.isImage(ocFile)) {
+                    mFiles.add(ocFile);
+                }
+
+                ContentValues cv = new ContentValues();
+                cv.put(ProviderMeta.ProviderTableMeta.VIRTUAL_TYPE, type.toString());
+                cv.put(ProviderMeta.ProviderTableMeta.VIRTUAL_OCFILE_ID, ocFile.getFileId());
+
+                contentValues.add(cv);
+            } catch (RemoteOperationFailedException e) {
+                Log_OC.e(TAG, "Error saving file with parent" + e.getMessage(),e);
             }
-
-            ContentValues cv = new ContentValues();
-            cv.put(ProviderMeta.ProviderTableMeta.VIRTUAL_TYPE, type.toString());
-            cv.put(ProviderMeta.ProviderTableMeta.VIRTUAL_OCFILE_ID, ocFile.getFileId());
-
-            contentValues.add(cv);
         }
 
         mStorageManager.saveVirtuals(type, contentValues);

--- a/src/modified/AndroidManifest.xml
+++ b/src/modified/AndroidManifest.xml
@@ -19,8 +19,8 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           xmlns:tools="http://schemas.android.com/tools"
           package="com.owncloud.android"
-          android:versionCode="20000004"
-          android:versionName="2.0.0RC4">
+          android:versionCode="20000005"
+          android:versionName="2.0.0RC5">
 
     <application
         android:name=".MainApp"

